pkg/option/config.go

```diff
@@ -2735,6 +2711,160 @@ func (c *DaemonConfig) populateHostServicesProtos() error {
+	}
+	if c.CTMapEntriesGlobalTCP > LimitTableMax || c.CTMapEntriesGlobalAny > LimitTableMax {
+		return fmt.Errorf("specified CT tables values %d/%d must not exceed maximum %d",
+			c.CTMapEntriesGlobalTCP, c.CTMapEntriesGlobalAny, LimitTableMax)
+	}
+
+	if c.NATMapEntriesGlobal < LimitTableMin {
+		return fmt.Errorf("specified NAT table size %d must exceed minimum %d",
+			c.NATMapEntriesGlobal, LimitTableMin)
+	}
+	if c.NATMapEntriesGlobal > LimitTableMax {
+		return fmt.Errorf("specified NAT tables size %d must not exceed maximum %d",
+			c.NATMapEntriesGlobal, LimitTableMax)
+	}
+	if c.NATMapEntriesGlobal > c.CTMapEntriesGlobalTCP+c.CTMapEntriesGlobalAny {
+		if c.NATMapEntriesGlobal == NATMapEntriesGlobalDefault {
+			// Auto-size for the case where CT table size was adapted but NAT still on default
+			c.NATMapEntriesGlobal = int((c.CTMapEntriesGlobalTCP + c.CTMapEntriesGlobalAny) * 2 / 3)
+		} else {
+			return fmt.Errorf("specified NAT tables size %d must not exceed maximum CT table size %d",
+				c.NATMapEntriesGlobal, c.CTMapEntriesGlobalTCP+c.CTMapEntriesGlobalAny)
+		}
+	}
+
+	if c.PolicyMapEntries < PolicyMapMin {
+		return fmt.Errorf("specified PolicyMap max entries %d must exceed minimum %d",
+			c.PolicyMapEntries, PolicyMapMin)
+	}
+	if c.PolicyMapEntries > PolicyMapMax {
+		return fmt.Errorf("specified PolicyMap max entries %d must not exceed maximum %d",
+			c.PolicyMapEntries, PolicyMapMax)
+	}
+
+	if c.FragmentsMapEntries < FragmentsMapMin {
+		return fmt.Errorf("specified fragments tracking map max entries %d must exceed minimum %d",
+			c.FragmentsMapEntries, FragmentsMapMin)
+	}
+	if c.FragmentsMapEntries > FragmentsMapMax {
+		return fmt.Errorf("specified fragments tracking map max entries %d must not exceed maximum %d",
+			c.FragmentsMapEntries, FragmentsMapMax)
+	}
+
+	return nil
+}
+
+func (c *DaemonConfig) calculateBPFMapSizes() error {
+	// BPF map size options
+	// Any map size explicitly set via option will override the dynamic
+	// sizing.
+	c.CTMapEntriesGlobalTCP = viper.GetInt(CTMapEntriesGlobalTCPName)
+	c.CTMapEntriesGlobalAny = viper.GetInt(CTMapEntriesGlobalAnyName)
+	c.NATMapEntriesGlobal = viper.GetInt(NATMapEntriesGlobalName)
+	c.PolicyMapEntries = viper.GetInt(PolicyMapEntriesName)
+
+	// Allow the range (0.0, 1.0] because the dynamic size will anyway be
+	// clamped to the table limits. Thus, a ratio of e.g. 0.98 will not lead
+	// to 98% of the total memory being allocated for BPF maps.
+	dynamicSizeRatio := viper.GetFloat64(MapEntriesGlobalDynamicSizeRatioName)
+	if 0.0 < dynamicSizeRatio && dynamicSizeRatio <= 1.0 {
+		vms, err := mem.VirtualMemory()
+		if err != nil || vms == nil {
+			log.WithError(err).Fatal("Failed to get system memory")
+		}
+		c.calculateDynamicBPFMapSizes(vms.Total, dynamicSizeRatio)
+	} else if dynamicSizeRatio < 0.0 {
+		return fmt.Errorf("specified dynamic map size ratio %f must be ≥ 0.0", dynamicSizeRatio)
+	}
```

Comment:
What happens if `dynamicSizeRatio > 1.0`? No error message?

Reply:
Good point. The comparison above is ` 0.0 < dynamicSizeRatio && dynamicSizeRatio <= 1.0` so a `dynamicSizeRatio > 1.0` will lead to an incorrect error message. Will adjust the error message.